### PR TITLE
feat(wysiwyg): Add Markdown support

### PR DESCRIPTION
**Pull request by Hywan**
This PR adds a new trait: `ToMardown` behind a new `to-markdown` Cargo feature. It allows to transform a DOM tree into Markdown.

In the test suite, by default, `assert_to_md` will roundtrip from HTML to Markdown, to HTML again. Otherwise, I've used `assert_to_md_no_roundtrip`. It happens when the Markdown to HTML compiler isn't able to parse our generated Markdown (it seems to be a bug, because it works with other parsers…) or because there is some formatting added (like newlines). This “roundtrip guarantee” isn't necessary for now. I was anticipating a next feature :-).

## Progress

- [x] line break node
- [x] text node
- [x] container node
  - [x] generic node
  - [x] formatting node
    - [x] bold
    - [x] italic
    - [x] strike through
    - [x] underline
    - [x] inline code 
  - [x] link node
  - [x] list node
  - [x] list item node

### CODE REVIEW

File: crates/wysiwyg/src/dom/to_markdown.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use super::UnicodeString;
+use bitflags::bitflags;
+use std::{error::Error, fmt};
+
+#[derive(Debug)]
+pub enum MarkdownError<S>
+where
+    S: UnicodeString,
+{
+    InvalidListItem(Option<<S::Str as ToOwned>::Owned>),
```
Comment (jplatte) [inline crates/wysiwyg/src/dom/to_markdown.rs:24]:
```suggestion
    InvalidListItem(Option<S>),
```

Comment (Hywan) [reply crates/wysiwyg/src/dom/to_markdown.rs:24]: True!

File: crates/wysiwyg/src/dom/nodes/container_node.rs [hunk 2]
```diff
@@ -374,6 +381,265 @@ where
...
+
+                child.fmt_markdown(buffer, options)?;
+            }
+
+            Ok(())
+        }
+
+        let mut options = *options;
+
+        match self.kind() {
```
Comment (andybalaam) [inline crates/wysiwyg/src/dom/nodes/container_node.rs:421]: This got long. What do you think about breaking the match arms into separate functions?

File: crates/wysiwyg/Cargo.toml [hunk 3]
```diff
@@ -10,13 +10,17 @@ name = "wysiwyg"
...
 once_cell = "1.13.0"
 widestring = "1.0.2"
 
 [dev-dependencies]
 speculoos = "0.9"
+pulldown-cmark = { version = "0.9.2", default-features = false }
+
+[features]
+default = ["to-markdown"]
+to-markdown = ["dep:bitflags"]
```
Comment (andybalaam) [inline crates/wysiwyg/Cargo.toml:26]: I'm not sure I see the benefit of breaking out a to-markdown feature - we are always going to need it to send Matrix messages.

File: crates/wysiwyg/src/dom/to_markdown.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    fn to_markdown(&self) -> Result<S, MarkdownError<S>> {
+        let mut buffer = S::default();
+        self.fmt_markdown(&mut buffer, &MarkdownOptions::empty())?;
+
+        Ok(buffer)
+    }
+}
+
+bitflags! {
```
Comment (andybalaam) [inline crates/wysiwyg/src/dom/to_markdown.rs:60]: Is it worth this dependency just to hold some (currently 1!) boolean flag?

File: crates/wysiwyg/src/tests/test_to_markdown.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    assert_to_md(
+        "abc<br /><br />def",
+        r#"abc\
+\
+def"#,
+    );
+}
+
+#[test]
+fn text_with_italic() {
```
Comment (andybalaam) [inline crates/wysiwyg/src/tests/test_to_markdown.rs:46]: Great tests!

File: crates/wysiwyg/src/tests/test_to_markdown.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+line2\
+\
+line3* def"#,
+    );
+}
+
+#[test]
+fn text_with_bold() {
+    assert_to_md("<strong>abc</strong>", "__abc__");
+    assert_to_md_no_roundtrip("abc<strong>def</strong>ghi", "abc__def__ghi");
```
Comment (andybalaam) [inline crates/wysiwyg/src/tests/test_to_markdown.rs:64]: Assuming this markdown is valid, we could consider fixing this as a bug in our markdown library (or choosing another that works better?)

File: crates/wysiwyg/src/tests/test_to_markdown.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    assert_to_md("abc <code>def</code> ghi", "abc `` def `` ghi");
+    assert_to_md("abc<code> def </code>ghi", "abc``  def  ``ghi");
+    // It's impossible to get a line break inside an inline code with Markdown.
+    assert_to_md_no_roundtrip(
+        "abc <code>line1<br />line2<br /><br />line3</code> def",
+        "abc `` line1 line2  line3 `` def",
+    );
+    assert_to_md_no_roundtrip(
+        "abc <code>def <strong>ghi</strong> jkl</code> mno",
+        "abc `` def __ghi__ jkl `` mno",
```
Comment (andybalaam) [inline crates/wysiwyg/src/tests/test_to_markdown.rs:126]: Is this valid markdown? Maybe we should strip formatting inside code blocks?

File: crates/wysiwyg/src/tests/test_to_markdown.rs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+}
+
+#[test]
+fn list_ordered() {
+    assert_to_md(
+        r#"<ol><li>item1</li><li>item2</li></ol>"#,
+        r#"1. item1
+2. item2"#,
+    );
+}
```
Comment (andybalaam) [inline crates/wysiwyg/src/tests/test_to_markdown.rs:168]: I'd like to see some tests with multiple nested ordered and unordered lists, with formatting and code blocks inside them.

File: crates/wysiwyg/src/tests/test_to_markdown.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    use md_parser::{html, Options, Parser};
+
+    let mut options = Options::empty();
+    options.insert(Options::ENABLE_STRIKETHROUGH);
+
+    let markdown = markdown.as_ustr().to_string_lossy();
+
+    let parser = Parser::new_ext(&markdown, options);
+
+    let mut html = String::new();
```
Comment (andybalaam) [inline crates/wysiwyg/src/tests/test_to_markdown.rs:201]: This confused me having a variable called `html` as well as `md_parser::html` so close together. Can we think of a different name for the variable?

**Review by andybalaam — CHANGES_REQUESTED** [review]
This is great, thank you. Some questions, minor comments and one new test.

File: crates/wysiwyg/src/dom/nodes/container_node.rs (hunk 2 above)
Comment (Hywan) [reply crates/wysiwyg/src/dom/nodes/container_node.rs:421]: Functions will be super small, and the `match` will stay. What's long is the comments, not the code. If you want me to change, that's fine, but I'm not sure it gonna make things any simpler: more code (due to function declaration), and more jump in the code.

File: crates/wysiwyg/Cargo.toml (hunk 3 above)
Comment (Hywan) [reply crates/wysiwyg/Cargo.toml:26]: I thought a feature was a good idea for the moment since it's still experimental. But I'm fine removing it. Tell me what you would prefer :-).

File: crates/wysiwyg/src/dom/to_markdown.rs (hunk 4 above)
Comment (Hywan) [reply crates/wysiwyg/src/dom/to_markdown.rs:60]: Haha, probably not! Previously I got 3 options, but we can remove `bitflags` for sure ;-).

File: crates/wysiwyg/src/tests/test_to_markdown.rs (hunk 6 above)
Comment (Hywan) [reply crates/wysiwyg/src/tests/test_to_markdown.rs:64]: To clarify, the problem doesn't come from “our” code, but from `pulldown` I guess. The problem is that Markdown is not very clear about what to do here. I'll see with `pulldown` why they did this as such.

File: crates/wysiwyg/src/tests/test_to_markdown.rs (hunk 7 above)
Comment (Hywan) [reply crates/wysiwyg/src/tests/test_to_markdown.rs:126]: So… yeah, having 2 backticks + 1 space as a surroundding pair is valid! See [the specification](https://github.github.com/gfm/#example-339). Using 2 backticks allows to have one backtick inside the code. It's the most weird-syntax-proof syntax I can come with. Example, to write  `` `foo _bar_ baz` `` I need to write:

```md
`` `foo _bar_ baz` ``
```

But why can't ensure a roundtrip here?

The HTML code is valid. The resulting Markdown is valid. But the specification is ambiguous here.

The following HTML code `<code>foo <em>bar</em> baz</code>` must map to `` `foo _bar_ baz` `` in Markdown. That's correct. However, `` `foo _bar_ baz` `` doesn't necessarily map to `<code>foo <em>bar</em> baz</code>` back to HTML. Actually, let's see what the Github compiler does: “`foo _bar_ baz`”. It renders this `foo \_bar\_ baz` (without the backslash). The `_` isn't compiled to `<em>`. Annoying right? That's because HTML allows inline element inside a `<code>` element, but Markdown doesn't.

File: crates/wysiwyg/src/tests/test_to_markdown.rs (hunk 8 above)
Comment (Hywan) [reply crates/wysiwyg/src/tests/test_to_markdown.rs:168]: OK!

File: crates/wysiwyg/src/tests/test_to_markdown.rs (hunk 9 above)
Comment (Hywan) [reply crates/wysiwyg/src/tests/test_to_markdown.rs:201]: Sure.

File: crates/wysiwyg/src/dom/nodes/container_node.rs (hunk 2 above)
Comment (andybalaam) [reply crates/wysiwyg/src/dom/nodes/container_node.rs:421]: Yeah, I do want it broken up please. Long functions, even if it's comments making them long, make it hard to follow the flow.

File: crates/wysiwyg/Cargo.toml (hunk 3 above)
Comment (andybalaam) [reply crates/wysiwyg/Cargo.toml:26]: Yeah, remove it please.

File: crates/wysiwyg/src/dom/to_markdown.rs (hunk 4 above)
Comment (andybalaam) [reply crates/wysiwyg/src/dom/to_markdown.rs:60]: Great, please do.

File: crates/wysiwyg/src/tests/test_to_markdown.rs (hunk 6 above)
Comment (andybalaam) [reply crates/wysiwyg/src/tests/test_to_markdown.rs:64]: Yeah, just musing here that we could consider reporting/fixing a bug in pulldown.

File: crates/wysiwyg/src/tests/test_to_markdown.rs (hunk 7 above)
Comment (andybalaam) [reply crates/wysiwyg/src/tests/test_to_markdown.rs:126]: Yeah, which is fine: we know that markdown does not support all of HTML. So arguably the right behaviour is to render

```
abc <code>def <strong>ghi</strong> jkl</code> mno
```

as

```
abc `` def ghi jkl `` mno
```

but it's a corner case and not something we need to do right now.

File: crates/wysiwyg/src/dom/to_markdown.rs (hunk 4 above)
Comment (Hywan) [reply crates/wysiwyg/src/dom/to_markdown.rs]: Done.

File: crates/wysiwyg/src/tests/test_to_markdown.rs (hunk 8 above)
Comment (Hywan) [reply crates/wysiwyg/src/tests/test_to_markdown.rs:187]: Done.

File: crates/wysiwyg/src/tests/test_to_markdown.rs (hunk 9 above)
Comment (Hywan) [reply crates/wysiwyg/src/tests/test_to_markdown.rs:220]: Done.

File: crates/wysiwyg/src/dom/nodes/container_node.rs (hunk 2 above)
Comment (Hywan) [reply crates/wysiwyg/src/dom/nodes/container_node.rs]: Done, https://github.com/matrix-org/matrix-wysiwyg/pull/143/commits/568250f784fa180fbc0343b3e40e821349fce322.

**Comment by Hywan** [comment]
All feedbacks have been addressed :-).

**Review by andybalaam — APPROVED** [review]

**Comment by andybalaam** [comment]
Awesome, thank you!
